hr_governance/models/governance_circle.py

```diff
@@ -0,0 +1,638 @@
+        forbidden_keys = ["expectation", "authority", "purpose"]
+        return forbidden_keys
+
+    def _get_hierarchy_records(self, reverse=False, include_self=False):
+        """Return the hiearchy records in ``self`` based on reverse direction.
+
+        :param reverse: If True, returns parent records.
+        By default, return child records
+        """
+        result = self.browse(
+            set.union(
+                set(), *self._get_role_ids_per_circle_id(reverse=reverse).values()
+            )
+        )
+        if include_self:
+            return self | result
+        return result
+
+    def _get_role_ids_per_circle_id(self, reverse):
+        return {rec.id: rec._get_roles_recursively(reverse=reverse).ids for rec in self}
+
+    def _get_roles_recursively(self, reverse):
+        roles = self.parent_id if reverse else self.child_ids
+        if not roles:
+            return self.env["governance.circle"]
+        return roles + roles._get_roles_recursively(reverse)
+
+    def unlink(self):
+        # Add roles to batch of records to delete
+        records = self._get_hierarchy_records(include_self=True)
+        # sorted to process child first
+        self = records.sorted(key=lambda x: x.id, reverse=True)
+        for rec in self:
+            if rec.member_rel_ids:
+                rec.member_rel_ids.unlink()
+        return super().unlink()
+
+    @api.model
+    def js_get_deleted_circle_info(self, circle_id):
+        result = {}
+        if circle_id:
+            circle = self.browse(circle_id)
+            children = circle._get_hierarchy_records()
+            result.update(
+                {
+                    "subcircles": len(children.filtered("is_circle").ids),
+                    "roles": len(children.filtered(lambda x: not x.is_circle).ids),
+                    "employees": len(children.member_rel_ids.ids),
+                }
+            )
+        return result
+
+    def _is_assigned(self):
+        self.ensure_one()
+        return (
+            self.is_circle
+            and self.child_ids.search_count(
+                [
+                    ("parent_id", "=", self.id),
+                    ("type_id.enable_edit_circle", "=", True),
+                    ("member_rel_ids", "!=", False),
+                ],
+            )
+            > 0
+        )
+
+    def _is_assigned_w_steering(self):
+        self.ensure_one()
+        steering_role = self.env.ref(
+            "hr_governance.steering_gct", raise_if_not_found=False
+        )
+        return (
+            steering_role
+            and self.is_circle
+            and self.child_ids.search_count(
+                [
+                    ("parent_id", "=", self.id),
+                    ("type_id", "=", steering_role.id),
+                    ("member_rel_ids", "!=", False),
+                ],
+            )
+            > 0
+        )
+
+    @api.model
+    def _get_steering_role_user_ids(self, circle_id):
```

Comment:
I don't think it makes sense to have this method as `@api.model`.

```suggestion
    def _get_steering_role_user_ids(self):
```

And, instead of calling the method with `model._get_steering_role_user_ids(circle)`, you'd call with `circle._get_steering_role_user_ids()`